### ISSUE TITLE
Reproducibility

### ISSUE THREAD

**Issue by jordan-schneider** [astooke/rlpyt#136]
I'm having issues with reproducibility. I run the following basic setup code twice, with fixed seed=0. The code saves the agent model after initialization and again after one step of training. Neither the initialization nor the post-training models are the same between runs. This is despite MinibatchRl.startup() setting the seed using rlpyt.utils.seed.set_seed. Nothing is running on GPU, so this setup should be deterministic, yes? Please let me know if you are aware of any underlying nondeterminism in torch that I'm not aware of, or if this is a real bug/unintended behavior. Models are compared both using diff on the saved files, and using torch.load and manually inspecting model weights (they're not even close).

```python
import torch

from rlpyt.agents.qpg.sac_agent import SacAgent
from rlpyt.algos.qpg.sac import SAC
from rlpyt.envs.gym import make as gym_make
from rlpyt.runners.minibatch_rl import MinibatchRl
from rlpyt.samplers.serial.sampler import SerialSampler

def main():
    seed = 0
    env_name = "MountainCarContinuous-v0"
    batch_T = 4
    batch_B = 1

    agent = SacAgent()

    algo_kwargs = {
        "OptimCls": torch.optim.SGD,
        "optim_kwargs": {"nesterov": True, "momentum": 0.9, "weight_decay": 0},
        "learning_rate": 0.9,
        "min_steps_learn": 1,
        "batch_size": batch_T,
        "replay_ratio": 1,
    }
    algo = SAC(**algo_kwargs)

    sampler_kwargs = {
        "EnvCls": gym_make,
        "env_kwargs": {"id": env_name},
        "eval_env_kwargs": {"id": env_name},
        "batch_T": batch_T,
        "batch_B": batch_B,
        "max_decorrelation_steps": 0,
    }
    sampler = SerialSampler(**sampler_kwargs)

    runner = MinibatchRl(
        agent=agent,
        algo=algo,
        sampler=sampler,
        n_steps=batch_T,
        log_interval_steps=batch_T,
        seed=seed,
    )
    runner.startup()
    torch.save(agent.state_dict(), "init_dict_1") # I replace this with "init_dict_2" on the second run.

    runner.train()
    torch.save(agent.state_dict(), "state_dict_1")

if __name__ == "__main__":
    main()
```

**Comment by jordan-schneider** [comment]
I have confirmed that torch.manual_seed produced deterministic initializations on my system.

**Comment by jordan-schneider** [comment]
And similarly for numpy and random.

**Comment by jordan-schneider** [comment]
I can't find where the gym env seeds are set, if anywhere, so that might be one bug (https://harald.co/2019/07/30/reproducibility-issues-using-openai-gym/), but I'm not sure why that would cause initialization unreproducability.

**Comment by jordan-schneider** [comment]
Adding
```python
        if seed is None:
            seed = make_seed()
        for env in envs:
            env.seed(seed)
            env.action_space.seed(seed)
            env.observation_space.seed(seed)
```
on line 45 of rlpyt.samplers.serial.sampler.py to set the gym environment seeds did not cause either to be deterministic. There is still definitely something in the initialization that's nondeterministic. (This fix also required adding a seed method to rlpyt.spaces.gym_wrapper.GymSpaceWrapper which just passed the call to the GymSpaceWrapper.seed method.)

**Comment by jordan-schneider** [comment]
The following does produce deterministic initializations:
```python
import gym
import torch
from rlpyt.agents.qpg.sac_agent import SacAgent
from rlpyt.envs.gym import GymEnvWrapper

env_name = "MountainCarContinuous-v0"
env = GymWrapper(gym.make(env_name))
env.seed(0)
torch.manual_seed(0)
agent = SacAgent()
agent.initialize(env.spaces)
torch.save(agent.state_dict, "state_dict")
```
So whatever is causing the non-determinism isn't in the gym wrapper or SacAgent. The sampler initializes the agent, so I'll try that next.

**Comment by jordan-schneider** [comment]
Ok now I'm actually getting determinism for both. Removing the env fix gets determinism for only the initial dicts. I'll submit a PR for the gym env seeding.

**Comment by astooke** [comment]
And finished with 8f36bda2ddc57c5d828eea0d69569f325cb8dc88
